test_harness/benchmarks/categories/src/hebench_benchmark_category.cpp

```diff
@@ -103,78 +103,95 @@ bool PartialBenchmarkCategory::validateResult(IDataLoader::Ptr dataset,
-            throw std::invalid_argument(IL_LOG_MSG_CLASS("Unexpected null output component in: 'outputs'."));
-
-        if (outputs.front()->size < truths.front()->size)
-            throw std::invalid_argument(IL_LOG_MSG_CLASS("Buffer in outputs is not large enough to contain the expected output: 'outputs'."));
-
-        std::uint64_t count = truths.front()->size / IDataLoader::sizeOf(data_type);
-        void *p_truth       = truths.front()->p;
-        void *p_output      = outputs.front()->p; // single output
-
-        // validate the results
-        switch (data_type)
+        std::vector<const hebench::APIBridge::NativeDataBuffer *>::iterator it_truths = truths.begin();
+        if (*it_truths)
         {
-        case hebench::APIBridge::DataType::Int32:
-            is_valid = hebench::Utilities::Math::almostEqual(reinterpret_cast<const std::int32_t *>(p_truth),
-                                                             reinterpret_cast<const std::int32_t *>(p_output),
-                                                             count,
-                                                             0.01);
-            break;
-
-        case hebench::APIBridge::DataType::Int64:
-            is_valid = hebench::Utilities::Math::almostEqual(reinterpret_cast<const std::int64_t *>(p_truth),
-                                                             reinterpret_cast<const std::int64_t *>(p_output),
-                                                             count,
-                                                             0.01);
-            break;
-
-        case hebench::APIBridge::DataType::Float32:
-            is_valid = hebench::Utilities::Math::almostEqual(reinterpret_cast<const float *>(p_truth),
-                                                             reinterpret_cast<const float *>(p_output),
-                                                             count,
-                                                             0.01);
-            break;
-
-        case hebench::APIBridge::DataType::Float64:
-            is_valid = hebench::Utilities::Math::almostEqual(reinterpret_cast<const double *>(p_truth),
-                                                             reinterpret_cast<const double *>(p_output),
-                                                             count,
-                                                             0.01);
-            break;
-
-        default:
-            retval = false;
-            break;
-        } // end switch
-
-        retval = retval && is_valid.empty();
-
-        if (!retval)
-        {
-            std::stringstream ss;
-            ss << "Elements not within 1% of each other, " << is_valid.size() << std::endl
-               << "Failed indices, ";
-            for (std::size_t i = 0; i < is_valid.size() && i < MaxErrorPrint; ++i)
+            for (auto it_outputs = outputs.begin(); retval && it_truths != truths.end(); ++it_truths, ++it_outputs)
```

Comment:
This works... but I'd think it would be better if we use index instead of iterators. It may make the code a bit clearer, and help with error message (see next comment).